languages/elixir/exercises/concept/agents/lib/community_garden.ex

```diff
@@ -0,0 +1,45 @@
+defmodule Plot do
+  @enforce_keys [:plot_id, :registered_to]
+  defstruct [:plot_id, :registered_to]
+end
+
+defmodule CommunityGarden do
+  def start(opts \\ []) do
+    {initial_registry, opts} = Keyword.pop(opts, :initial_registry, %{})
+    {initial_next_id, opts} = Keyword.pop(opts, :initial_next_id, 1)
+    Agent.start(fn -> %{registry: initial_registry, next_id: initial_next_id} end, opts)
+  end
+
+  def list_registrations(pid) do
+    Agent.get(pid, fn state ->
+      state.registry
+    end)
+  end
+
+  def register(pid, register_to) do
+    Agent.get_and_update(pid, fn %{registry: registry, next_id: next_id} = state ->
+      new_plot = %Plot{plot_id: next_id, registered_to: register_to}
+      updated = Map.put(registry, next_id, new_plot)
+      {new_plot, %{state | registry: updated, next_id: next_id + 1}}
+    end)
+  end
+
+  def release(pid, plot_id) do
+    Agent.update(pid, fn %{registry: registry} = state ->
+      updated = Map.delete(registry, plot_id)
+      %{state | registry: updated}
+    end)
+  end
+
+  def get_registration(pid, plot_id) do
+    registration =
+      Agent.get(pid, fn state ->
+        state.registry[plot_id]
+      end)
+
+    case registration do
+      nil -> {:error, "plot is unregistered"}
+      _ -> registration
+    end
```

Comment:
Conceptually, is it really an error to ask if a certain plot is registered when it isn't? I think I would just leave out this case expression, it doesn't add much to the topic at hand and a `nil` return value makes perfect sense to me in this case.

Reply:
I can agree that `:error` should be something more meaningful/relevant (perhaps `:not_found` is an improvement). But I think that returning nil doesn't really tell you why you didn't get back a result. Was it not found, did it time out, was the key an invalid type?

I think the same argument could be applied here, it doesn't add much but it does provide some caveat which requires the student to think about the domain of the story rather than just copying Agent functions which return the state.